pytest_inmanta/plugin.py

```diff
@@ -1358,6 +1392,143 @@ def finalize_all_handlers(self) -> None:
+        :param run_as_root: run the handler as root or not
+        :param dry_run: only perform dryrun
+        :param status: expected status after deploy,
+            set to None to not check,
+            for deploy defaults to deployed
+            for dryrun defaults to dry
+        """
+        if expected_status == DEFAULT:
+            expected_status = (
+                const.ResourceState.deployed if not dry_run else const.ResourceState.dry
+            )
+
+        resource = self.get_resource(resource_type, **filter_args)
+        assert resource is not None, "No resource found of given type and filter args"
+
+        h = self.get_handler(resource, run_as_root)
+        assert h is not None
+
+        ctx = handler.HandlerContext(resource, dry_run=dry_run)
+        h.execute(ctx, resource, dry_run)
+        self.finalize_context(ctx)
+        self.finalize_handler(h)
+
+        out = DeployResultV2(resource=resource, ctx=ctx, handler=h)
+        if expected_status is not None:
+            out.assert_status(expected_status)
+
+        return out
+
+    def dryrun_resource_v2(
+        self,
+        resource_type: str,
+        run_as_root: bool = False,
+        expected_status: Optional[const.ResourceState] = DEFAULT,
+        **filter_args: object,
+    ) -> "DeployResultV2":
+        return self.deploy_resource_v2(
+            resource_type, run_as_root, True, expected_status, **filter_args
+        )
+
+
+@dataclass
+class DeployResultV2:
+    ctx: HandlerContext
+    handler: ResourceHandler
+    resource: Resource
+
+    # status
+    def assert_status(
+        self,
+        status: const.ResourceState = const.ResourceState.deployed,
+        change: const.Change = None,
+    ):
+        ctx = self.ctx
+        if ctx.status != status:
+            print("Deploy did not result in correct status")
+            print("Requested changes: ", ctx._changes)
+            for log in ctx.logs:
+                print("Log: ", log._data["msg"])
+                print(
+                    "Kwargs: ",
+                    [
+                        "%s: %s" % (k, v)
+                        for k, v in log._data["kwargs"].items()
+                        if k != "traceback"
+                    ],
+                )
+                if "traceback" in log._data["kwargs"]:
+                    print("Traceback:\n", log._data["kwargs"]["traceback"])
+
+        assert ctx.status == status
+        if change is not None:
+            assert ctx._change == change
+        self.assert_consistent_status()
+
+    # Discovery
+    @property
+    def discovered_resources(self) -> list[object]:
+        return self.handler._client.discovered_resources
+
+    # Logs
+    @property
+    def logs(self) -> list[LogLine]:
+        return self.ctx.logs
+
+    def assert_has_logline(self, matches: str) -> LogLine:
+        """
+         Assert the handler logged a log line matching the pattern
+
+        :return: that logline
+        """
```

Comment:
It could be made more clear that a regex is expected here.  Either by taking as input the pattern object, or documenting it clearly